middle_end/flambda/types/env/aliases.ml

```diff
@@ -257,27 +380,74 @@ let empty = {
 }
 
 type canonical =
-  | Is_canonical of Simple.t
-  | Alias_of_canonical of { element : Simple.t; canonical_element : Simple.t; }
+  | Is_canonical
+  | Alias_of_canonical of {
+      canonical_element : Simple.t;
+      coercion_to_canonical : coercion_to_canonical;
+    }
 
 let canonical t element : canonical =
   match Simple.Map.find element t.canonical_elements with
-  | exception Not_found -> Is_canonical element
-  | canonical_element ->
+  | exception Not_found -> Is_canonical
+  | canonical_element, coercion_to_canonical ->
     if !Clflags.flambda_invariant_checks then begin
       assert (not (Simple.equal element canonical_element))
     end;
-    Alias_of_canonical { element; canonical_element; }
+    Alias_of_canonical { canonical_element; coercion_to_canonical; }
 
 let get_aliases_of_canonical_element t ~canonical_element =
   match Simple.Map.find canonical_element t.aliases_of_canonical_elements with
   | exception Not_found -> Aliases_of_canonical_element.empty
   | aliases -> aliases
 
-let add_alias_between_canonical_elements t ~canonical_element ~to_be_demoted =
-  if Simple.equal canonical_element to_be_demoted then
-    t
-  else
+(*
+   before
+   ~~~~~~
+   canonical_element <--[coercion_ce_0]-- ce_0
+     ^ ^--[...]-- ...
+      \--[coercion_ce_m]-- ce_m
+   to_be_demoted <--[coercion_tbd_0]-- tbd_0
+     ^ ^--[...]-- ...
+      \--[coercion_tbd_n]-- tbd_n
+
+   i.e.
+
+   canonical_elements[ce_0] = (canonical_element, coercion_ce_0)
+   ...
+   canonical_elements[ce_m] = (canonical_element, coercion_ce_m)
+   canonical_elements[tbd_0] = (to_be_demoted, coercion_tbd_0)
+   ...
+   canonical_elements[tbd_n] = (to_be_demoted, coercion_tbd_n)
+
+
+   after
+   ~~~~~
+   canonical_element <--[coercion_ce_0]-- ce_0
+     ^ ^ ^ ^ ^ ^--[...]-- ...
+     | | | |  \--[coercion_ce_m]-- ce_m
+     | | | \--[coercion_to_canonical]-- to_be_demoted
+     | | \--[compose(coercion_tbd_0, coercion_to_canonical)]-- tbd_0
+     | \--[...]-- ...
+     \--[compose(coercion_tbd_n, coercion_to_canonical)]-- tbd_n
+
+   i.e.
+
+   canonical_elements[ce_0] = (canonical_element, coercion_ce_0)
+   ...
+   canonical_elements[ce_m] = (canonical_element, coercion_ce_m)
+   canonical_elements[to_be_demoted] = (canonical_element, coercion_to_canonical)
+   canonical_elements[tbd_0] = (canonical_element, compose(coercion_tbd_0, coercion_to_canonical))
+   ...
+   canonical_elements[tbd_n] = (canonical_element, compose(coercion_tbd_n, coercion_to_canonical))
+
+*)
+let add_alias_between_canonical_elements t ~canonical_element ~coercion_to_canonical:{ coercion_to_canonical; } ~to_be_demoted =
```

Comment:
I believe [coercion_to_canonical] doesn't need to be a record here. [add_alias_between_canonical_elements] is only called once (L559) and the record is manually constructed only to be unboxed here.

Reply:
There isn't actually any boxing or unboxing happening here since `coercion_to_canonical` carries the `@@unboxed` annotation. I think the `coercion_to_canonical` type is going away anyway (to avoid having to "unbox" whole maps, for one thing).